jgrapht-core/src/main/java/org/jgrapht/alg/connectivity/DulmageMendelsohnDecomposition.java

```diff
@@ -0,0 +1,303 @@
+/*
+ * (C) Copyright 2018-2018, by CAE Tech Limited and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.connectivity;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.jgrapht.Graph;
+import org.jgrapht.alg.interfaces.MatchingAlgorithm.Matching;
+import org.jgrapht.alg.interfaces.StrongConnectivityAlgorithm;
+import org.jgrapht.alg.matching.HopcroftKarpMaximumCardinalityBipartiteMatching;
+import org.jgrapht.graph.AsSubgraph;
+import org.jgrapht.graph.DefaultDirectedGraph;
+import org.jgrapht.graph.DefaultEdge;
+import org.jgrapht.graph.EdgeReversedGraph;
+import org.jgrapht.graph.builder.GraphBuilder;
+import org.jgrapht.traverse.DepthFirstIterator;
+
+/**
+ * The Dulmage-Mendelsohn Decomposition
```

Comment:
Also, for the complexity, it would be good to note that if fine decomposition is requested, then the ceiling is higher due to the Hopcroft-Karp matching.